File: Online_Judge_Problems/boj/boj_11866_요세푸스_문제0/boj_11866.c
/*-----------------------------------------------------
Sub  : [BOJ] 요세푸스 문제0
Link : https://www.acmicpc.net/problem/11866
Level: 실버 4
Tag  : C, Queue
Memo
-----------------------------------------------------*/

#include <stdio.h>
#include <stdlib.h>

int main() {
    int N, K;
    scanf("%d %d", &N, &K);

    int *arr = (int *)malloc(N * sizeof(int));
    for (int i = 0; i < N; i++) arr[i] = i + 1;

    int size = N, idx = 0;
    printf("<");

    for (int i = 0; i < N; i++) {
        idx = (idx + K - 1) % size;  // K번째 사람
        printf("%d", arr[idx]);

        if (i != N - 1) printf(", ");

        for (int j = idx; j < size - 1; j++) {
            arr[j] = arr[j + 1];  // 제거 후 시프트
        }
        size--;
    }
    printf(">\n");

    free(arr);
    return 0;
}

File: boj/boj_11866_요세푸스_문제0/boj_11866.js
/*-----------------------------------------------------
Sub  : [BOJ] 요세푸스 문제0
Link : https://www.acmicpc.net/problem/11866
Level: 실버 4
Tag  : JS, Queue
Memo
-----------------------------------------------------*/

const TEST_MODE = true;

let input;

if (TEST_MODE) {
  const filePath = require('path').join(__dirname, 'input.txt');
  input = require('fs').readFileSync(filePath, 'utf-8').trim().split('\n');
}
else {
  input = require('fs').readFileSync(0, 'utf-8').trim().split('\n');
}


// 입력
const [N, K] = input[0]

// 풀이 함수
function josephus(N, K) {
    let q = [];
    for (let i = 1; i <= N; i++) q.push(i);

    let result = [];
    let idx = 0;

    while (q.length > 0) {
        idx = (idx + K - 1) % q.length; // K번째 이동
        result.push(q.splice(idx, 1)[0]); // 제거
    }
    return result;
}

// 출력
let result = josephus(N, K);
console.log("<" + result.join(", ") + ">");
